##### Load libraries ###############################################################################
library("tidyr")
library("dplyr")
library("purrr")
library("broom")
library("ipred")
library("pec")
library("GEOquery")
#### Load data #######################################################################################
load("data.Rdata")
##### Define constants ######################################################################################
datasets = c("ExprImageRaw", "ExprImageFil", "METABRIC", "GSE11121", "GSE19615", "GSE22226A", "GSE22226B", "GSE26971", 
             "GSE96058", "GSE1456A","GSE1456B", "GSE42568", "GSE45255", "GSE4922A", "GSE4922B", "GSE7390", 
             "GSE7849", "GSE9195", "GSE9893", "GSE2603", "NKI", "TCGA")
##### Merge clinical and expression ######################################################################################
df = tibble(dataset = datasets,
            clinical_data = list(
              data.ExprImage.raw,
              data.ExprImage.fil,
              data.MB.clin,
              data.X1.clin,
              data.X2.clin,
              data.X3A.clin,
              data.X3B.clin,
              data.X4.clin,
              data.X5.clin,
              data.X6A.clin,
              data.X6B.clin,
              data.X7.clin,
              data.X8.clin,
              data.X9A.clin,
              data.X9B.clin,
              data.X10.clin,
              data.X11.clin,
              data.X12.clin,
              data.X13.clin,
              data.X14.clin,
              data.X15.clin,
              data.X16.clin),
            expression_data = list(
              data.ExprImage.raw,
              data.ExprImage.fil,
              data.MB.expr,
              data.X1.expr,
              data.X2.expr,
              data.X3A.expr,
              data.X3B.expr,
              data.X4.expr,
              data.X5.expr,
              data.X6A.expr,
              data.X6B.expr,
              data.X7.expr,
              data.X8.expr,
              data.X9A.expr,
              data.X9B.expr,
              data.X10.expr,
              data.X11.expr,
              data.X12.expr,
              data.X13.expr,
              data.X14.expr,
              data.X15.expr,
              data.X16.expr)
            )

#### Get experiment data #####################################################################################
gse_datasets = unlist(map(df$clinical_data, function(x){return(x$study_id[1])})) [4:19]

get_experiment_data = function(gse_id){
  print(paste0("Loading ", gse_id ))
  G = getGEO(gse_id, GSEMatrix = TRUE)
  return(experimentData(G[[1]]))
}

experiment_data_gse = map(gse_datasets , get_experiment_data) 
experiment_data_exprimage = paste("Background: The receptors for estrogen (ESR1) and progesterone (PGR) are both part of the same signaling pathway and routinely used for breast cancer stratification. We tested the hypothesis if a coordinated analysis could add extra information for prognostic stratification. Materials and methods: ESR1 and PGR gene expression was first investigated by quantitative reverse transcription polymerase chain reaction in fresh-frozen invasive ductal breast cancer samples (Hamburg collective, case-control, n=317). Our results were then tested using two datasets generated by different technical approaches: i) a public DNA-chip data set (GSE3494, n=251) and ii) semiquantitative protein expression data based on immunohistochemistry (Stuttgart collective, n=18,528). Results: The PGR/ESR1 gene-expression ratio was a prognostic indicator in those with ESR1/PGR-positive breast cancer (Hamburg collective), with a high PGR/ESR1 expression ratio indicating a favorable outcome. In all three collectives, the PGR/ESR1 mRNA ratio or its protein equivalent was a univariate prognostic factor and also a multivariate prognostic factor in the Hamburg and Stuttgart collectives. Conclusion: Calculation of the PGR/ESR1 gene-expression ratio and its immunohistochemical surrogate could be a useful and simple addition to routine breast cancer diagnostics. A high PGR/ESR1 ratio could be indicative of a favorable clinical outcome.")
experiment_data_metabric = paste("The elucidation of breast cancer subgroups and their molecular drivers requires integrated views of the genome and transcriptome from representative numbers of patients. We present an integrated analysis of copy number and gene expression in a discovery and validation set of 997 and 995 primary breast tumours, respectively, with long-term clinical follow-up. Inherited variants (CNVs, SNPs) and acquired somatic copy number aberrations (CNAs) were associated with expression in 40% of genes, although the landscape was dominated by cis and trans-acting CNAs. By delineating expression outlier genes driven in cis by CNAs, we identified putative cancer genes, including deletions in PPP2R2A, MTAP, and MAP2K4. Unsupervised analysis of paired DNA/RNA profiles revealed novel subgroups with distinct clinical outcomes, which reproduced in the validation cohort. These include a high-risk, ER-positive 11q13/14 cis-acting subgroup and a favourable prognosis subgroup devoid of CNAs. Trans-acting aberration 0152hotspots were found to modulate subgroup-specific gene networks, including a TCR deletion-mediated adaptive immune response in the 0152CNA-devoid sub-group and a Basal-specific chromosome 5 deletion-driven mitotic network. Our results provide a novel molecular stratification of the breast cancer population, derived from the impact of somatic copy number aberrations on the transcriptome")
experiment_data_nki = paste("Background: A more accurate means of prognostication in breast cancer will improve the selection of patients for adjuvant systemic therapy. Methods: Using microarray analysis to evaluate our previously established 70-gene prognosis profile, we classified a series of 295 consecutive patients with primary breast carcinomas as having a gene-expression signature associated with either a poor prognosis or a good prognosis. All patients had stage I or II breast cancer and were younger than 53 years old; 151 had lymph-node-negative disease, and 144 had lymph-node-positive disease. We evaluated the predictive power of the prognosis profile using univariable and multivariable statistical analyses. Results: Among the 295 patients, 180 had a poor-prognosis signature and 115 had a good-prognosis signature, and the mean (+/-SE) overall 10-year survival rates were 54.6+/-4.4 percent and 94.5+/-2.6 percent, respectively. At 10 years, the probability of remaining free of distant metastases was 50.6+/-4.5 percent in the group with a poor-prognosis signature and 85.2+/-4.3 percent in the group with a good-prognosis signature. The estimated hazard ratio for distant metastases in the group with a poor-prognosis signature, as compared with the group with the good-prognosis signature, was 5.1 (95 percent confidence interval, 2.9 to 9.0; P<0.001). This ratio remained significant when the groups were analyzed according to lymph-node status. Multivariable Cox regression analysis showed that the prognosis profile was a strong independent factor in predicting disease outcome. Conclusions: The gene-expression profile we studied is a more powerful predictor of the outcome of disease in young patients with breast cancer than standard systems based on clinical and histologic criteria.")
experiment_data_tcga = paste("We analyzed primary breast cancers by genomic DNA copy number arrays, DNA methylation, exome sequencing, mRNA arrays, microRNA sequencing and reverse phase protein arrays. Our ability to integrate information across platforms provided key insights into previously-defined gene expression subtypes and demonstrated the existence of four main breast cancer classes when combining data from five platforms, each of which shows significant molecular heterogeneity. Somatic mutations in only three genes (TP53, PIK3CA and GATA3) occurred at > 10% incidence across all breast cancers; however, there were numerous subtype-associated and novel gene mutations including the enrichment of specific mutations in GATA3, PIK3CA and MAP3K1 with the Luminal A subtype. We identified two novel protein expression-defined subgroups, possibly contributed by stromal/microenvironmental elements, and integrated analyses identified specific signaling pathways dominant in each molecular subtype including a HER2/p-HER2/HER1/p-HER1 signature within the HER2-Enriched expression subtype. Comparison of Basal-like breast tumors with high-grade Serous Ovarian tumors showed many molecular commonalities, suggesting a related etiology and similar therapeutic opportunities. The biologic finding of the four main breast cancer subtypes caused by different subsets of genetic and epigenetic abnormalities raises the hypothesis that much of the clinically observable plasticity and heterogeneity occurs within, and not across, these major biologic subtypes of breast cancer.")

experiment_data = c(experiment_data_exprimage, experiment_data_exprimage, experiment_data_metabric, 
                    experiment_data_gse, experiment_data_nki, experiment_data_tcga)

df = df %>% 
  mutate(experiment_data = experiment_data)

#### Get titles ################################################################################
title = c("Improved Risk Stratification for Breast Cancer Samples Based on the Expression Ratio of the Estrogen and Progesterone Receptor",
          "Improved Risk Stratification for Breast Cancer Samples Based on the Expression Ratio of the Estrogen and Progesterone Receptor",
          "Associations between genomic stratification of breast cancer and centrally reviewed tumour pathology in the METABRIC cohort",
          map(experiment_data_gse, function(x) return(x@title[[1]])),
          "A gene-expression signature as a predictor of survival in breast cancer",
          "Comprehensive molecular portraits of human breast tumours")

df = df %>% 
  mutate(title = title)





























